**<h2> Manipulando Dados em DataFrames do Pandas</h2>**

In [87]:
# cria um dicionário

dados = {'Estado': ['Santa Catarina','Rio de Janeiro','Tocantins', 'Bahia','Minas Gerais'],
         'Ano':[2004,2005,2006,2007,2008],
         'Taxa Desemprego':[1.5,1.7,1.6,2.4,2.7]
         }

In [88]:
import pandas as pd

In [89]:
from pandas import DataFrame

In [90]:
df = DataFrame(dados)

In [91]:
#Visualiza as 5 primeiras linhas
df.head()

Estado   Ano  Taxa Desemprego
0  Santa Catarina  2004              1.5
1  Rio de Janeiro  2005              1.7
2       Tocantins  2006              1.6
3           Bahia  2007              2.4
4    Minas Gerais  2008              2.7

In [92]:
type(df)

pandas.core.frame.DataFrame

In [93]:
#Reorganizando as colunas

DataFrame(dados,columns=['Estado','Taxa Desemprego','Ano'])

Estado  Taxa Desemprego   Ano
0  Santa Catarina              1.5  2004
1  Rio de Janeiro              1.7  2005
2       Tocantins              1.6  2006
3           Bahia              2.4  2007
4    Minas Gerais              2.7  2008

In [94]:
#Criando outro dataframe com os mesmos dados anteriores mas adicionando uma coluna

df2 = DataFrame(dados,columns=['Estado','Taxa Desemprego','Taxa Crescimento','Ano'],
                               index = ['estado1','estado2','estado3','estado4','estado5'])

In [95]:
df2

Estado  Taxa Desemprego Taxa Crescimento   Ano
estado1  Santa Catarina              1.5              NaN  2004
estado2  Rio de Janeiro              1.7              NaN  2005
estado3       Tocantins              1.6              NaN  2006
estado4           Bahia              2.4              NaN  2007
estado5    Minas Gerais              2.7              NaN  2008

In [96]:
df2.values

array([['Santa Catarina', 1.5, nan, 2004],
       ['Rio de Janeiro', 1.7, nan, 2005],
       ['Tocantins', 1.6, nan, 2006],
       ['Bahia', 2.4, nan, 2007],
       ['Minas Gerais', 2.7, nan, 2008]], dtype=object)

In [97]:
df2.dtypes

Estado               object
Taxa Desemprego     float64
Taxa Crescimento     object
Ano                   int64
dtype: object

In [98]:
df2.columns

Index(['Estado', 'Taxa Desemprego', 'Taxa Crescimento', 'Ano'], dtype='object')

In [99]:
# Imprimindo apenas uma coluna do DataFrame
df2['Estado']

estado1    Santa Catarina
estado2    Rio de Janeiro
estado3         Tocantins
estado4             Bahia
estado5      Minas Gerais
Name: Estado, dtype: object

In [100]:
#Imprimindo apenas duas colunas do DataFrame
df2[['Taxa Desemprego','Ano']]

Taxa Desemprego   Ano
estado1              1.5  2004
estado2              1.7  2005
estado3              1.6  2006
estado4              2.4  2007
estado5              2.7  2008

In [101]:
#Filtrando pelo índice

df2.filter(items=['estado3'],axis = 0)

Estado  Taxa Desemprego Taxa Crescimento   Ano
estado3  Tocantins              1.6              NaN  2006

**<h2> Usando NumPy e Pandas para manipulação de  Dados</h2>**

In [102]:
df2.head()

Estado  Taxa Desemprego Taxa Crescimento   Ano
estado1  Santa Catarina              1.5              NaN  2004
estado2  Rio de Janeiro              1.7              NaN  2005
estado3       Tocantins              1.6              NaN  2006
estado4           Bahia              2.4              NaN  2007
estado5    Minas Gerais              2.7              NaN  2008

In [103]:
#Resumo estatístico do DataFrame
df2.describe()

Taxa Desemprego          Ano
count         5.000000     5.000000
mean          1.980000  2006.000000
std           0.535724     1.581139
min           1.500000  2004.000000
25%           1.600000  2005.000000
50%           1.700000  2006.000000
75%           2.400000  2007.000000
max           2.700000  2008.000000

In [104]:
#Verifica se tem valor NaN
df2.isna()

Estado  Taxa Desemprego  Taxa Crescimento    Ano
estado1   False            False              True  False
estado2   False            False              True  False
estado3   False            False              True  False
estado4   False            False              True  False
estado5   False            False              True  False

In [105]:
df2['Taxa Crescimento'].isna()

estado1    True
estado2    True
estado3    True
estado4    True
estado5    True
Name: Taxa Crescimento, dtype: bool

In [106]:
import numpy as np

In [107]:
df2['Taxa Crescimento'] = np.arange(5.)

In [108]:
df2

Estado  Taxa Desemprego  Taxa Crescimento   Ano
estado1  Santa Catarina              1.5               0.0  2004
estado2  Rio de Janeiro              1.7               1.0  2005
estado3       Tocantins              1.6               2.0  2006
estado4           Bahia              2.4               3.0  2007
estado5    Minas Gerais              2.7               4.0  2008

In [109]:
df2['Taxa Crescimento'].isna()

estado1    False
estado2    False
estado3    False
estado4    False
estado5    False
Name: Taxa Crescimento, dtype: bool

In [110]:
df2.describe()

Taxa Desemprego  Taxa Crescimento          Ano
count         5.000000          5.000000     5.000000
mean          1.980000          2.000000  2006.000000
std           0.535724          1.581139     1.581139
min           1.500000          0.000000  2004.000000
25%           1.600000          1.000000  2005.000000
50%           1.700000          2.000000  2006.000000
75%           2.400000          3.000000  2007.000000
max           2.700000          4.000000  2008.000000

**<h2> Slicing de DataFrame do Pandas</h2>**

In [111]:
df2

Estado  Taxa Desemprego  Taxa Crescimento   Ano
estado1  Santa Catarina              1.5               0.0  2004
estado2  Rio de Janeiro              1.7               1.0  2005
estado3       Tocantins              1.6               2.0  2006
estado4           Bahia              2.4               3.0  2007
estado5    Minas Gerais              2.7               4.0  2008

In [112]:
df2['estado2':'estado4']

Estado  Taxa Desemprego  Taxa Crescimento   Ano
estado2  Rio de Janeiro              1.7               1.0  2005
estado3       Tocantins              1.6               2.0  2006
estado4           Bahia              2.4               3.0  2007

In [113]:
df2[df2['Taxa Desemprego']< 2]

Estado  Taxa Desemprego  Taxa Crescimento   Ano
estado1  Santa Catarina              1.5               0.0  2004
estado2  Rio de Janeiro              1.7               1.0  2005
estado3       Tocantins              1.6               2.0  2006

In [114]:
df2[['Estado','Taxa Crescimento']]

Estado  Taxa Crescimento
estado1  Santa Catarina               0.0
estado2  Rio de Janeiro               1.0
estado3       Tocantins               2.0
estado4           Bahia               3.0
estado5    Minas Gerais               4.0

In [115]:
df2[['Estado','Taxa Crescimento','Ano']]

Estado  Taxa Crescimento   Ano
estado1  Santa Catarina               0.0  2004
estado2  Rio de Janeiro               1.0  2005
estado3       Tocantins               2.0  2006
estado4           Bahia               3.0  2007
estado5    Minas Gerais               4.0  2008

**<h2> Preenchendo Valores Ausentes em DataFrames do Pandas</h2>**

A função fillna() é usada para preencher os valores ausentes. A função oferece muitas opções. Podemos usar um valor específico, uma função agregada(por exemplo, média) ou o valore anterior ou seguinte.

Para esse exemplo usaremos a moda, a estatística que representa o valor que aparece mais vezes em uma variável.

In [116]:
# Primeiro importamos um dataset
dsa_df = pd.read_csv('dataset.csv')

In [117]:
dsa_df.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         NaN  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         NaN  
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0

In [118]:
dsa_df.isna().sum()

ID_Pedido       0
Data_Pedido     0
ID_Cliente      0
Segmento        0
Pais            0
Regiao          0
ID_Produto      0
Categoria       0
Nome_Produto    0
Valor_Venda     0
Quantidade      2
dtype: int64

In [119]:
# Extraímos a moda da coluna Quantity
moda = dsa_df['Quantidade'].value_counts().index[0]

In [120]:
moda

3.0

In [121]:
# E por fim preenchemos os valores NA com a moda
dsa_df['Quantidade'].fillna(value=moda,inplace=True)

#inplace = True é como se fosse pra salvar o resultado dentro do dataframe original

In [122]:
dsa_df.isna().sum()

ID_Pedido       0
Data_Pedido     0
ID_Cliente      0
Segmento        0
Pais            0
Regiao          0
ID_Produto      0
Categoria       0
Nome_Produto    0
Valor_Venda     0
Quantidade      0
dtype: int64

**<h2> Query(Consulta) de Dados no DataFrame do Pandas</h2>**

Com o Pandas criamos dataframes, que são essencialmente tabelas.Como tal, podemos fazer consultas, ou simplesmente queries. E para isso usamos o método query(). Veja o exemplo abaixo:

In [123]:
#Checamos os valores mínimo e máximo da coluna Valor_Venda
dsa_df['Valor_Venda'].describe()

count     9994.000000
mean       229.858001
std        623.245101
min          0.444000
25%         17.280000
50%         54.490000
75%        209.940000
max      22638.480000
Name: Valor_Venda, dtype: float64

O intervalo de Vendas (Valor_Venda) é de 0.44 a 22638. Vamos fazer uma consulta e retornar todas as vendas dentro de um range de valores. Fazemos isso com a instrução abaixo

In [124]:
#Geramos um novo dataframe apenas com o intervalo proposto
df2 = dsa_df.query('229 < Valor_Venda < 10000')

In [125]:
df2

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
0     CA-2016-152156  2016-11-08   CG-12520     Consumer  United States   
1     CA-2016-152156  2016-11-08   CG-12520     Consumer  United States   
3     US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
7     CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
10    CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
...              ...         ...        ...          ...            ...   
9973  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9976  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9979  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9991  CA-2017-121258  2017-02-26   DB-13060     Consumer  United States   
9993  CA-2017-119914  2017-05-04   CC-12220     Consumer  United States   

     Regiao       ID_Produto        Categoria  \
0     South  FUR-BO-10001798        Furniture   
1     South  FUR-CH-10000454        Furniture   
3     South  FUR-TA-10000577        Furniture   
7      West  TEC-PH-10002275       Technology   
10     West  FUR-TA-10001539        Furniture   
...     ...              ...              ...   
9973   West  TEC-PH-10004080       Technology   
9976   West  TEC-PH-10002496       Technology   
9979   West  OFF-BI-10002026  Office Supplies   
9991   West  TEC-PH-10003645       Technology   
9993   West  OFF-AP-10002684  Office Supplies   

                                           Nome_Produto  Valor_Venda  \
0                     Bush Somerset Collection Bookcase     261.9600   
1     Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400   
3         Bretford CR4500 Series Slim Rectangular Table     957.5775   
7                        Mitel 5320 IP Phone VoIP phone     907.1520   
10             Chromcraft Rectangular Conference Tables    1706.1840   
...                                                 ...          ...   
9973                           Avaya 5410 Digital phone     271.9600   
9976                                       Cisco SPA301     249.5840   
9979                  Ibico Recycled Linen-Style Covers     437.4720   
9991                              Aastra 57i VoIP phone     258.5760   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...     243.1600   

      Quantidade  
0            3.0  
1            3.0  
3            5.0  
7            6.0  
10           9.0  
...          ...  
9973         5.0  
9976         2.0  
9979        14.0  
9991         2.0  
9993         2.0  

[2357 rows x 11 columns]

In [126]:
df2['Valor_Venda'].describe()

count    2357.000000
mean      766.679142
std       856.315136
min       229.544000
25%       323.100000
50%       490.320000
75%       859.200000
max      9892.740000
Name: Valor_Venda, dtype: float64

**<h2>Verificando a Ocorrência de Diversos Valores em uma Coluna</h2>**

Em nosso conjunto de dados de exemplo, temos a coluna Quantidade que representa a qtde de itens vendidos em cada uma das vendas. Imagine que precisamos saber em quais vendas foram vendidos 5,7,9 ou 11 itens.

Como aplicaríamos esse filtro ao nosso dataframe?

Fácil. O Pandas oferece o método isin() para checar diversos valores em uma coluna. Quem conhece SQL já deve ter percebido que é o método IN em SQL. ex

In [127]:
dsa_df.shape

(9994, 11)

In [128]:
# Então aplicamos o filtro

dsa_df[dsa_df['Quantidade'].isin([5,7,9,11])]

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
3     US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
5     CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
9     CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
10    CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
14    US-2015-118983  2015-11-22   HP-14815  Home Office  United States   
...              ...         ...        ...          ...            ...   
9974  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9977  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9981  CA-2017-163566  2017-08-03   TB-21055     Consumer  United States   
9982  US-2016-157728  2016-09-22   RC-19960     Consumer  United States   
9988  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   

       Regiao       ID_Produto        Categoria  \
3       South  FUR-TA-10000577        Furniture   
5        West  FUR-FU-10001487        Furniture   
9        West  OFF-AP-10002892  Office Supplies   
10       West  FUR-TA-10001539        Furniture   
14    Central  OFF-AP-10002311  Office Supplies   
...       ...              ...              ...   
9974     West  OFF-AR-10004752  Office Supplies   
9977     West  OFF-FA-10003467  Office Supplies   
9981     East  OFF-LA-10004484  Office Supplies   
9982  Central  OFF-PA-10002195  Office Supplies   
9988    South  TEC-PH-10004006       Technology   

                                           Nome_Produto  Valor_Venda  \
3         Bretford CR4500 Series Slim Rectangular Table     957.5775   
5     Eldon Expressions Wood and Plastic Desk Access...      48.8600   
9                      Belkin F5C206VTEL 6 Outlet Surge     114.9000   
10             Chromcraft Rectangular Conference Tables    1706.1840   
14    Holmes Replacement Filter for HEPA Air Cleaner...      68.8100   
...                                                 ...          ...   
9974                               Blackstonian Pencils      18.6900   
9977           Alliance Big Bands Rubber Bands, 12/Pack      13.8600   
9981                                          Avery 476      16.5200   
9982  RSVP Cards & Envelopes, Blank White, 8-1/2" X ...      35.5600   
9988                    Panasonic KX - TS880B Telephone     206.1000   

      Quantidade  
3            5.0  
5            7.0  
9            5.0  
10           9.0  
14           5.0  
...          ...  
9974         7.0  
9977         7.0  
9981         5.0  
9982         7.0  
9988         5.0  

[2128 rows x 11 columns]

Na instrução acima estamos filtrando o dataframe df, retornando as linhas onde a coluna Quantidade for igual aos valores 5,7,9 ou 11. Passamos uma lista de valores como argumento para o método isin()

Vamos deixar um pouco mais divertido. Vamos retornar somente 10 linhas, não 2128:

In [129]:
dsa_df[dsa_df['Quantidade'].isin([5,7,9,11])].shape

(2128, 11)

In [130]:
dsa_df[dsa_df['Quantidade'].isin([5,7,9,11])][:10]

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
3   US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
5   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
9   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
10  CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
14  US-2015-118983  2015-11-22   HP-14815  Home Office  United States   
21  CA-2016-137330  2016-12-09   KB-16585    Corporate  United States   
22  CA-2016-137330  2016-12-09   KB-16585    Corporate  United States   
27  US-2015-150630  2015-09-17   TB-21520     Consumer  United States   
35  CA-2016-117590  2016-12-08   GH-14485    Corporate  United States   
36  CA-2016-117590  2016-12-08   GH-14485    Corporate  United States   

     Regiao       ID_Produto        Categoria  \
3     South  FUR-TA-10000577        Furniture   
5      West  FUR-FU-10001487        Furniture   
9      West  OFF-AP-10002892  Office Supplies   
10     West  FUR-TA-10001539        Furniture   
14  Central  OFF-AP-10002311  Office Supplies   
21  Central  OFF-AR-10000246  Office Supplies   
22  Central  OFF-AP-10001492  Office Supplies   
27     East  FUR-BO-10004834        Furniture   
35  Central  TEC-PH-10004977       Technology   
36  Central  FUR-FU-10003664        Furniture   

                                         Nome_Produto  Valor_Venda  Quantidade  
3       Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
5   Eldon Expressions Wood and Plastic Desk Access...      48.8600         7.0  
9                    Belkin F5C206VTEL 6 Outlet Surge     114.9000         5.0  
10           Chromcraft Rectangular Conference Tables    1706.1840         9.0  
14  Holmes Replacement Filter for HEPA Air Cleaner...      68.8100         5.0  
21                                         Newell 318      19.4600         7.0  
22        Acco Six-Outlet Power Strip, 4' Cord Length      60.3400         7.0  
27  Riverside Palais Royal Lawyers Bookcase, Royal...    3083.4300         7.0  
35                                        GE 30524EE4    1097.5440         7.0  
36  Electrix Architect's Clamp-On Swing Arm Lamp, ...     190.9200         5.0

**<h2> Operadores Lógicos para Manipulação de Dados com Pandas</h2>**

In [131]:
# Filtrando as vendas que ocorreram para o segmento de home office e na região South
dsa_df[ (dsa_df['Segmento'] == 'Home Office') & (dsa_df['Regiao']=='South')].head()

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais Regiao  \
182  CA-2014-158274  2014-11-19   RM-19675  Home Office  United States  South   
183  CA-2014-158274  2014-11-19   RM-19675  Home Office  United States  South   
184  CA-2014-158274  2014-11-19   RM-19675  Home Office  United States  South   
231  US-2017-100930  2017-04-07   CS-12400  Home Office  United States  South   
232  US-2017-100930  2017-04-07   CS-12400  Home Office  United States  South   

          ID_Produto   Categoria  \
182  TEC-PH-10003273  Technology   
183  TEC-PH-10004896  Technology   
184  TEC-AC-10002345  Technology   
231  FUR-TA-10001705   Furniture   
232  FUR-TA-10003473   Furniture   

                                         Nome_Produto  Valor_Venda  Quantidade  
182                                      AT&T TR1909W     503.9600         4.0  
183                        Nokia Lumia 521 (T-Mobile)     149.9500         5.0  
184                 HP Standard 104 key PS/2 Keyboard      29.0000         2.0  
231  Bush Advantage Collection Round Conference Table     233.8600         2.0  
232        Bretford Rectangular Conference Table Tops     620.6145         3.0

In [132]:
# Usando operador OR
dsa_df[(dsa_df['Segmento'] == 'Home Office') | (dsa_df['Regiao']=='South')]

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
0     CA-2016-152156  2016-11-08   CG-12520     Consumer  United States   
1     CA-2016-152156  2016-11-08   CG-12520     Consumer  United States   
3     US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
4     US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
12    CA-2017-114412  2017-04-15   AA-10480     Consumer  United States   
...              ...         ...        ...          ...            ...   
9979  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9980  US-2015-151435  2015-09-06   SW-20455     Consumer  United States   
9987  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   
9988  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   
9989  CA-2014-110422  2014-01-21   TB-21400     Consumer  United States   

     Regiao       ID_Produto        Categoria  \
0     South  FUR-BO-10001798        Furniture   
1     South  FUR-CH-10000454        Furniture   
3     South  FUR-TA-10000577        Furniture   
4     South  OFF-ST-10000760  Office Supplies   
12    South  OFF-PA-10002365  Office Supplies   
...     ...              ...              ...   
9979   West  OFF-BI-10002026  Office Supplies   
9980  South  FUR-TA-10001029        Furniture   
9987  South  TEC-AC-10001539       Technology   
9988  South  TEC-PH-10004006       Technology   
9989  South  FUR-FU-10001889        Furniture   

                                           Nome_Produto  Valor_Venda  \
0                     Bush Somerset Collection Bookcase     261.9600   
1     Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400   
3         Bretford CR4500 Series Slim Rectangular Table     957.5775   
4                        Eldon Fold 'N Roll Cart System      22.3680   
12                                           Xerox 1967      15.5520   
...                                                 ...          ...   
9979                  Ibico Recycled Linen-Style Covers     437.4720   
9980                         KI Adjustable-Height Table      85.9800   
9987  Logitech G430 Surround Sound Gaming Headset wi...      79.9900   
9988                    Panasonic KX - TS880B Telephone     206.1000   
9989                             Ultra Door Pull Handle      25.2480   

      Quantidade  
0            3.0  
1            3.0  
3            5.0  
4            2.0  
12           3.0  
...          ...  
9979        14.0  
9980         1.0  
9987         1.0  
9988         5.0  
9989         3.0  

[3131 rows x 11 columns]

In [133]:
dsa_df[(dsa_df['Segmento'] == 'Home Office') | (dsa_df['Regiao']=='South')].tail()
#tail visualiza a parte final do dataframe

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
9979  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9980  US-2015-151435  2015-09-06   SW-20455     Consumer  United States   
9987  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   
9988  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   
9989  CA-2014-110422  2014-01-21   TB-21400     Consumer  United States   

     Regiao       ID_Produto        Categoria  \
9979   West  OFF-BI-10002026  Office Supplies   
9980  South  FUR-TA-10001029        Furniture   
9987  South  TEC-AC-10001539       Technology   
9988  South  TEC-PH-10004006       Technology   
9989  South  FUR-FU-10001889        Furniture   

                                           Nome_Produto  Valor_Venda  \
9979                  Ibico Recycled Linen-Style Covers      437.472   
9980                         KI Adjustable-Height Table       85.980   
9987  Logitech G430 Surround Sound Gaming Headset wi...       79.990   
9988                    Panasonic KX - TS880B Telephone      206.100   
9989                             Ultra Door Pull Handle       25.248   

      Quantidade  
9979        14.0  
9980         1.0  
9987         1.0  
9988         5.0  
9989         3.0

In [134]:
dsa_df[(dsa_df['Segmento'] != 'Home Office') & (dsa_df['Regiao']!='South')].sample(5)
#sample retorna uma amostra de 5 valores

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
7593  CA-2016-158778  2016-09-05   DB-13210   Consumer  United States   East   
4198  US-2016-161683  2016-12-03   CJ-12010   Consumer  United States   East   
5579  CA-2017-137498  2017-09-09   LC-17050   Consumer  United States   West   
8041  CA-2017-121489  2017-08-21   CM-11815  Corporate  United States   West   
8884  CA-2014-107398  2014-08-25   BS-11590  Corporate  United States   East   

           ID_Produto        Categoria  \
7593  FUR-CH-10000309        Furniture   
4198  OFF-BI-10001072  Office Supplies   
5579  FUR-CH-10003833        Furniture   
8041  TEC-AC-10000682       Technology   
8884  OFF-BI-10001982  Office Supplies   

                                           Nome_Produto  Valor_Venda  \
7593                    Global Comet Stacking Arm Chair      887.271   
4198  GBC Clear Cover, 8-1/2 x 11, unpunched, 25 cov...       18.192   
5579                          Novimex Fabric Task Chair      243.920   
8041  Kensington K72356US Mouse-in-a-Box USB Desktop...       33.180   
8884         Wilson Jones Custom Binder Spines & Labels        6.528   

      Quantidade  
7593         3.0  
4198         4.0  
5579         5.0  
8041         2.0  
8884         4.0

**<h2> Agrupamento de Dados em DataFrames com Group By</h2>**

A função Pandas Groupby é uma função versátil e fácil de usar que ajuda a obter uma visão geral dos dados. Isso torna mais fácil explorar o conjunto de dados e revelar os relacionamentos entre as variáveis.

O código a seguir irá agrupar as linhas com base nas combinações Segmento/Regiao/Valor_Venda e nos dará a taxa média de vendas de cada grupo

In [135]:
# Aplicamos o group by
dsa_df[['Segmento','Regiao','Valor_Venda']].groupby(['Segmento','Regiao']).mean()

Valor_Venda
Segmento    Regiao              
Consumer    Central   207.946728
            East      238.875539
            South     233.390180
            West      217.033955
Corporate   Central   234.763466
            East      228.516929
            South     238.992025
            West      235.265911
Home Office Central   208.248046
            East      253.911805
            South     272.996329
            West      239.442692

Na instrução acima primeiro filtramos os dados extraindo 3 colunas: [Segumento, Regiao, Valor_venda]. Na sequência, agrupamos por duas colunas:[Segmento,Regiao]. E então calculamos a média para a coluna que ficou fora do group by, nesse caso a coluna Valor_Venda

**<h2> Agregação Múltipla com GroupBy</ha>**

Vamos explorar mais a função groupby() pois temos diversas formas de sumarização dos dados de forma mais simples. No exemplo abaixo, uniremos a função groupby() com a função agg() para realizar agregação múltipla.

In [136]:
#Aplicamos o group by
dsa_df[['Segmento','Regiao','Valor_Venda']].groupby(['Segmento','Regiao']).agg(['mean','std','count'])

Valor_Venda                   
                           mean          std count
Segmento    Regiao                                
Consumer    Central  207.946728   587.906523  1212
            East     238.875539   633.371169  1469
            South    233.390180   559.346824   838
            West     217.033955   551.997547  1672
Corporate   Central  234.763466   818.947521   673
            East     228.516929   530.001654   877
            South    238.992025   586.176947   510
            West     235.265911   471.288764   960
Home Office Central  208.248046   371.009180   438
            East     253.911805   722.777318   502
            South    272.996329  1404.798466   272
            West     239.442692   529.242737   571

Na instrução acima primeiro filtramos os dados extraindo 3 colunas:[Segmento, Regiao, Valor_Venda]. Na sequência, agrupamos por duas colunas:[Segmento,Regiao]. E então agregamos os dados calculando a média, o desvio padrão e contagem de elementos para a coluna que ficou de fora do group by, no caso a coluna Valor_Venda.

A função agg() recebe como argumento uma lista de funções para agregação.

**<h2> Filtrando DataFrame do Pandas com Base em Strings</h2>**

In [137]:
dsa_df.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         3.0  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0  
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0

In [138]:
# Filtramos o dataframe pela coluna Segmento com valores que iniciam com as letras 'Con'
dsa_df[dsa_df['Segmento'].str.startswith('Con')].head()

ID_Pedido Data_Pedido ID_Cliente  Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520  Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520  Consumer  United States  South   
3  US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
5  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   
5  FUR-FU-10001487        Furniture   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         3.0  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0  
5  Eldon Expressions Wood and Plastic Desk Access...      48.8600         7.0

In [139]:
#Filtramos o dataframe pela coluna Segmento com valores que terminam com as letras 'mer'
dsa_df[dsa_df['Segmento'].str.endswith('mer')].head()

ID_Pedido Data_Pedido ID_Cliente  Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520  Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520  Consumer  United States  South   
3  US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
5  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   
5  FUR-FU-10001487        Furniture   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         3.0  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0  
5  Eldon Expressions Wood and Plastic Desk Access...      48.8600         7.0

As funções startswith() e endswith() são muito úteis quando for necessário filtrar strings por caracteres que apareçam no começo e/ou final

**<h2> Split de Strings em DataFrames do Pandas</h2>**

Com pandas podemos realizar diversas tarefas de split de strings dividindo uma coluna ou extraindo elementos do nosso interesse

In [140]:
dsa_df.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         3.0  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0  
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0

In [141]:
dsa_df['ID_Pedido'].head()

0    CA-2016-152156
1    CA-2016-152156
2    CA-2016-138688
3    US-2015-108966
4    US-2015-108966
Name: ID_Pedido, dtype: object

In [142]:
dsa_df['ID_Pedido'].str.split('-')

0       [CA, 2016, 152156]
1       [CA, 2016, 152156]
2       [CA, 2016, 138688]
3       [US, 2015, 108966]
4       [US, 2015, 108966]
               ...        
9989    [CA, 2014, 110422]
9990    [CA, 2017, 121258]
9991    [CA, 2017, 121258]
9992    [CA, 2017, 121258]
9993    [CA, 2017, 119914]
Name: ID_Pedido, Length: 9994, dtype: object

In [143]:
dsa_df['ID_Pedido'].str.split('-').str[1].head()

0    2016
1    2016
2    2016
3    2015
4    2015
Name: ID_Pedido, dtype: object

In [144]:
#Fazemos o split da coluna e extraímos o item na posição 2 (índice 1)
dsa_df['Ano'] = dsa_df['ID_Pedido'].str.split('-').str[1]

In [145]:
# Então conferimos a nova coluna criada
dsa_df.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  \
0                  Bush Somerset Collection Bookcase     261.9600         3.0   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0   
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0   
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0   
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0   

    Ano  
0  2016  
1  2016  
2  2016  
3  2015  
4  2015

**<h2> Strip de Strings em DataFrames do Pandas</h2>**

O split divide a string. O strip remove caracteres da string.

In [146]:
dsa_df.head(3)

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  \
0                  Bush Somerset Collection Bookcase       261.96         3.0   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...       731.94         3.0   
2  Self-Adhesive Address Labels for Typewriters b...        14.62         2.0   

    Ano  
0  2016  
1  2016  
2  2016

In [147]:
dsa_df['Data_Pedido'].head(3)

0    2016-11-08
1    2016-11-08
2    2016-06-12
Name: Data_Pedido, dtype: object

A coluna 'Data_Pedido' é a data de envio do produto no formato YYYY-MM-DD. Imagine que seja necessário deixar o ano apenas com 2 dígitos sem alterar o tipo da variável. Fazemos isso com a função lstrip(), ou seja, left strip.

In [148]:
#Vamos remover os dígitos 2 e 0 á esquerda do valor da variável 'Data_Pedido'
dsa_df['Data_Pedido'].str.lstrip('20')

0       16-11-08
1       16-11-08
2       16-06-12
3       15-10-11
4       15-10-11
          ...   
9989    14-01-21
9990    17-02-26
9991    17-02-26
9992    17-02-26
9993    17-05-04
Name: Data_Pedido, Length: 9994, dtype: object

Como não usamos o inplace = True, a mudança é somente na memória e não altera o dataframe. Podemos usar rstrip() e strip() com diferentes variações de strip de strings

**<h2> Replace de Strings em DataFrames do Pandas</h2>**

Se for necessário substituir caracteres dentro de uma string o Pandas oferece uma função (replace) para isso também.

In [149]:
dsa_df.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  \
0                  Bush Somerset Collection Bookcase     261.9600         3.0   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0   
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0   
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0   
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0   

    Ano  
0  2016  
1  2016  
2  2016  
3  2015  
4  2015

In [151]:
#Substituímos os caracteres CG por AX na coluna 'ID_Cliente'
dsa_df['ID_Cliente'] = dsa_df['ID_Cliente'].str.replace('CG','AX')

In [152]:
dsa_df.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   AX-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   AX-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  \
0                  Bush Somerset Collection Bookcase     261.9600         3.0   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0   
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0   
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0   
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0   

    Ano  
0  2016  
1  2016  
2  2016  
3  2015  
4  2015